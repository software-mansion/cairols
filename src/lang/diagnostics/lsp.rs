use std::collections::HashMap;

use cairo_lang_diagnostics::{
    DiagnosticEntry, DiagnosticLocation, Diagnostics, PluginFileDiagnosticNotes, Severity,
};
use cairo_lang_filesystem::db::FilesGroup;
use cairo_lang_filesystem::ids::FileId;
use cairo_lang_utils::Upcast;
use lsp_types::{
    Diagnostic, DiagnosticRelatedInformation, DiagnosticSeverity, Location, NumberOrString, Range,
    Url,
};
use tracing::{error, trace};

use crate::lang::lsp::{LsProtoGroup, ToLsp};

/// Converts internal diagnostics to LSP format.
pub fn map_cairo_diagnostics_to_lsp<'db, T>(
    db: &'db T::DbType,
    diags: &mut HashMap<(Url, FileId<'db>), Vec<Diagnostic>>,
    diagnostics: &Diagnostics<'db, T>,
    trace_macro_diagnostics: bool,
    plugin_file_notes: &PluginFileDiagnosticNotes<'db>,
) where
    T::DbType: salsa::Database + for<'a> Upcast<'a, dyn FilesGroup>,
    T: DiagnosticEntry<'db> + salsa::Update,
{
    for diagnostic in if trace_macro_diagnostics {
        diagnostics.get_all()
    } else {
        diagnostics.get_diagnostics_without_duplicates(db)
    } {
        let mut message = diagnostic.format(db);
        let diagnostic_location = diagnostic.location(db);
        let (_, parent_file_notes) =
            diagnostic_location.user_location_with_plugin_notes(db.upcast(), plugin_file_notes);
        let mut related_information = vec![];
        for note in diagnostic.notes(db).iter().chain(&parent_file_notes) {
            if let Some(location) = &note.location {
                let Some((range, file_id)) = get_mapped_range_and_add_mapping_note(
                    db,
                    location,
                    trace_macro_diagnostics.then_some(&mut related_information),
                    "Next note mapped from here.",
                ) else {
                    continue;
                };
                let Some(uri) = db.url_for_file(file_id) else {
                    trace!("url for file not found: {:?}", file_id.long(db));
                    continue;
                };
                related_information.push(DiagnosticRelatedInformation {
                    location: Location { uri, range },
                    message: note.text.clone(),
                });
            } else {
                message += &format!("\nnote: {}", note.text);
            }
        }

        let Some((range, mapped_file_id)) = get_mapped_range_and_add_mapping_note(
            db,
            &diagnostic.location(db),
            trace_macro_diagnostics.then_some(&mut related_information),
            "Diagnostic mapped from here.",
        ) else {
            continue;
        };

        let diagnostic = Diagnostic {
            range,
            message,
            related_information: (!related_information.is_empty()).then_some(related_information),
            severity: Some(match diagnostic.severity() {
                Severity::Error => DiagnosticSeverity::ERROR,
                Severity::Warning => DiagnosticSeverity::WARNING,
            }),
            code: diagnostic.error_code().map(|code| NumberOrString::String(code.to_string())),
            ..Diagnostic::default()
        };
        let Some(mapped_file_url) = db.url_for_file(mapped_file_id) else {
            continue;
        };

        diags.entry((mapped_file_url, mapped_file_id)).or_default().push(diagnostic);
    }
}

/// Returns the mapped range of a location, optionally adds a note about the mapping of the
/// location.
fn get_mapped_range_and_add_mapping_note<'db>(
    db: &'db (impl for<'a> Upcast<'a, dyn FilesGroup> + ?Sized),
    orig: &DiagnosticLocation<'db>,
    related_info: Option<&mut Vec<DiagnosticRelatedInformation>>,
    message: &str,
) -> Option<(Range, FileId<'db>)> {
    let mapped = orig.user_location(db.upcast());
    let mapped_range = get_lsp_range(db.upcast(), &mapped)?;
    if let Some(related_info) = related_info {
        if *orig != mapped {
            if let Some(range) = get_lsp_range(db.upcast(), orig) {
                related_info.push(DiagnosticRelatedInformation {
                    location: Location { uri: db.url_for_file(orig.file_id)?, range },
                    message: message.to_string(),
                });
            }
        }
    }
    Some((mapped_range, mapped.file_id))
}

/// Converts an internal diagnostic location to an LSP range.
fn get_lsp_range<'db>(
    db: &'db dyn FilesGroup,
    location: &DiagnosticLocation<'db>,
) -> Option<Range> {
    let Some(span) = location.span.position_in_file(db, location.file_id) else {
        error!("failed to get range for diagnostic");
        return None;
    };
    Some(span.to_lsp())
}
