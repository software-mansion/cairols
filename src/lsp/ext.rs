//! CairoLS extensions to the Language Server Protocol.

use std::path::PathBuf;

use lsp_types::notification::Notification;
use lsp_types::request::Request;
use lsp_types::{TextDocumentPositionParams, Url};
use serde::{Deserialize, Serialize};

/// Provides content of virtual file from the database.
pub(crate) struct ProvideVirtualFile;

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileRequest {
    pub uri: Url,
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
pub(crate) struct ProvideVirtualFileResponse {
    pub content: Option<String>,
}

impl Request for ProvideVirtualFile {
    type Params = ProvideVirtualFileRequest;
    type Result = ProvideVirtualFileResponse;
    const METHOD: &'static str = "vfs/provide";
}

/// Collects information about all Cairo crates that are currently being analyzed.
pub struct ViewAnalyzedCrates;

impl Request for ViewAnalyzedCrates {
    type Params = ();
    type Result = String;
    const METHOD: &'static str = "cairo/viewAnalyzedCrates";
}

/// Provides string with code after macros expansion.
pub struct ExpandMacro;

impl Request for ExpandMacro {
    type Params = TextDocumentPositionParams;
    type Result = Option<String>;
    const METHOD: &'static str = "cairo/expandMacro";
}

/// Notifies about corelib version mismatch.
#[derive(Debug)]
pub struct CorelibVersionMismatch;

impl Notification for CorelibVersionMismatch {
    type Params = String;
    const METHOD: &'static str = "cairo/corelib-version-mismatch";
}

/// Collects versions of LS and it's dependencies.
#[derive(Debug)]
pub struct ToolchainInfo;

#[derive(Serialize, Deserialize)]
pub struct PathAndVersion {
    pub path: PathBuf,
    pub version: String,
}

#[derive(Serialize, Deserialize)]
pub struct ToolchainInfoResponse {
    pub ls: PathAndVersion,
    pub scarb: Option<PathAndVersion>,
}

impl Request for ToolchainInfo {
    type Params = ();
    type Result = ToolchainInfoResponse;
    const METHOD: &'static str = "cairo/toolchainInfo";
}

#[cfg(feature = "testing")]
pub mod testing {
    use lsp_types::notification::Notification;

    /// Notifies about the end of project updating.
    #[derive(Debug)]
    pub struct ProjectUpdatingFinished;

    impl Notification for ProjectUpdatingFinished {
        type Params = ();
        const METHOD: &'static str = "cairo/projectUpdatingFinished";
    }

    /// Notifies about diagnostics generation which is beginning to calculate.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationStart;

    impl Notification for DiagnosticsCalculationStart {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationStart";
    }

    /// Notifies about diagnostics generation which ended calculating.
    #[derive(Debug)]
    pub struct DiagnosticsCalculationFinish;

    impl Notification for DiagnosticsCalculationFinish {
        type Params = ();
        const METHOD: &'static str = "cairo/diagnosticsCalculationFinish";
    }
}

#[derive(Debug)]
pub struct ScarbPathMissing {}

impl Notification for ScarbPathMissing {
    type Params = ();
    const METHOD: &'static str = "scarb/could-not-find-scarb-executable";
}

#[derive(Debug)]
pub struct ScarbResolvingStart {}

impl Notification for ScarbResolvingStart {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-start";
}

#[derive(Debug)]
pub struct ScarbResolvingFinish {}

impl Notification for ScarbResolvingFinish {
    type Params = ();
    const METHOD: &'static str = "scarb/resolving-finish";
}
