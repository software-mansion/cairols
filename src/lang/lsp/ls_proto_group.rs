use std::num::NonZeroU32;

use cairo_lang_filesystem::db::FilesGroup;
use cairo_lang_filesystem::ids::{FileId, FileLongId};
use cairo_lang_filesystem::span::TextSpan;
use cairo_lang_utils::Intern;
use cairo_lang_utils::Upcast;
use lsp_types::{Location, Url};
use salsa::Id;
use tracing::error;

use crate::lang::lsp::ToLsp;

#[cfg(test)]
#[path = "ls_proto_group_test.rs"]
mod test;

pub trait LsProtoGroup: for<'db> Upcast<'db, dyn FilesGroup> {
    /// Get a [`FileId`] from an [`Url`].
    ///
    /// Returns `None` on failure, and errors are logged.
    fn file_for_url<'db>(&'db self, uri: &Url) -> Option<FileId<'db>> {
        match uri.scheme() {
            "file" => uri
                .to_file_path()
                .inspect_err(|()| error!("invalid file url: {uri}"))
                .ok()
                .map(|path| FileLongId::OnDisk(path).intern(self.upcast())),
            "vfs" => uri
                .host_str()
                .or_else(|| {
                    error!("invalid vfs url, missing host string: {uri:?}");
                    None
                })?
                .parse::<NonZeroU32>()
                .inspect_err(|e| {
                    error!("invalid vfs url, host string is not a valid integer, {e}: {uri:?}")
                })
                .ok()
                .map(|id| unsafe { Id::from_u32(id.into()) })
                .map(FileId::from_intern_id),
            _ => {
                error!("invalid url, scheme is not supported by this language server: {uri:?}");
                None
            }
        }
    }

    /// Get the canonical [`Url`] for a [`FileId`].
    fn url_for_file<'db>(&'db self, file_id: FileId<'db>) -> Option<Url> {
        let vf = match file_id.long(self.upcast()) {
            FileLongId::OnDisk(path) => return Some(Url::from_file_path(path).unwrap()),
            FileLongId::Virtual(vf) => vf,
            FileLongId::External(id) => &self.upcast().try_ext_as_virtual(*id)?,
        };
        // NOTE: The URL is constructed using setters and path segments in order to
        //   url-encode any funky characters in parts that LS is not controlling.
        let mut url = Url::parse("vfs://").unwrap();
        url.set_host(Some(&file_id.as_intern_id().as_u32().to_string())).unwrap();
        url.path_segments_mut().unwrap().push(&format!("{}.cairo", vf.name));
        Some(url)
    }

    /// Converts a [`FileId`]-[`TextSpan`] pair into a [`Location`].
    fn lsp_location<'db>(&'db self, (file, span): (FileId<'db>, TextSpan)) -> Option<Location> {
        let db = self.upcast();
        let found_uri = db.url_for_file(file)?;
        let range = span.position_in_file(db, file)?.to_lsp();
        let location = Location { uri: found_uri, range };
        Some(location)
    }
}

impl<T> LsProtoGroup for T where T: for<'db> Upcast<'db, dyn FilesGroup> + ?Sized {}
