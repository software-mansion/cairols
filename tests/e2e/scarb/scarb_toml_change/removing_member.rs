use cairo_language_server::lsp;
use indoc::indoc;
use lsp_types::notification::DidChangeWatchedFiles;
use lsp_types::{DidChangeWatchedFilesParams, FileChangeType, FileEvent};
use similar::TextDiff;

use super::{AnalyzedCratesResult, caps};
use crate::support::normalize::normalize;
use crate::support::sandbox;

#[test]
fn test_removing_member() {
    let mut ls = sandbox! {
        files {
            "Scarb.toml" => indoc! (r#"
                [workspace]
                members = [
                    "a",
                    "b"
                ]
            "#),
            "a/Scarb.toml" => indoc! (r#"
                [package]
                name = "a"
                version = "0.1.0"
                edition = "2024_07"

                [dependencies]
                b = { path = "../b" }
            "#),
            "a/src/lib.cairo" => "",
            "b/Scarb.toml" => indoc! (r#"
                [package]
                name = "b"
                version = "0.1.0"
                edition = "2024_07"
            "#),
            "b/src/lib.cairo" => "",
        }
        client_capabilities = caps;
    };

    assert!(ls.open_and_wait_for_diagnostics("a/src/lib.cairo").is_empty());
    // Check if opening `a` triggers calculating diagnostics for `b`.
    assert!(ls.get_diagnostics_for_file("b/src/lib.cairo").is_empty());

    let analyzed_crates = ls.send_request::<lsp::ext::ViewAnalyzedCrates>(());
    let analyzed_crates = normalize(&ls, analyzed_crates);

    ls.edit_file(
        "Scarb.toml",
        indoc!(
            r#"
                [workspace]
                members = [
                    "a",
                ]
            "#
        ),
    );
    ls.send_notification::<DidChangeWatchedFiles>(DidChangeWatchedFilesParams {
        changes: vec![FileEvent { uri: ls.doc_id("Scarb.toml").uri, typ: FileChangeType::CHANGED }],
    });
    ls.wait_for_project_update();

    let analyzed_crates_after_member_removal = ls.send_request::<lsp::ext::ViewAnalyzedCrates>(());
    let analyzed_crates_after_member_removal = normalize(&ls, analyzed_crates_after_member_removal);

    let analyzed_crates_diff =
        TextDiff::from_lines(&analyzed_crates, &analyzed_crates_after_member_removal)
            .unified_diff()
            .context_radius(15)
            .to_string();

    insta::assert_toml_snapshot!(AnalyzedCratesResult { analyzed_crates, analyzed_crates_diff })
}
