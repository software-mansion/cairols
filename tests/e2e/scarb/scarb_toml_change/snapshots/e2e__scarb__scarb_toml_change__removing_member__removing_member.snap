---
source: tests/e2e/scarb/scarb_toml_change/removing_member.rs
description: "// → Scarb.toml\n[workspace]\nmembers = [\n    \"a\",\n    \"b\"\n]\n\n// → a/Scarb.toml\n[package]\nname = \"a\"\nversion = \"0.1.0\"\nedition = \"2024_07\"\n\n[dependencies]\nb = { path = \"../b\" }\n\n// → a/src/lib.cairo\n\n\n// → b/Scarb.toml\n[package]\nname = \"b\"\nversion = \"0.1.0\"\nedition = \"2024_07\"\n\n// → b/src/lib.cairo\n"
expression: "AnalyzedCratesResult { analyzed_crates, analyzed_crates_diff }"
---
analyzed_crates = '''
# Analyzed Crates
---
```json
{
  "name": "a",
  "source_paths": [
    "[ROOT]/a/src/lib.cairo"
  ],
  "settings": {
    "name": "a",
    "edition": "2024_07",
    "version": "0.1.0",
    "cfg_set": [
      [
        "target",
        "lib"
      ],
      [
        "target",
        "test"
      ],
      "test"
    ],
    "dependencies": {
      "a": {
        "discriminator": "a 0.1.0 (path+[ROOT_URL]a/Scarb.toml)"
      },
      "b": {
        "discriminator": "b 0.1.0 (path+[ROOT_URL]b/Scarb.toml)"
      },
      "core": {
        "discriminator": null
      }
    },
    "experimental_features": {
      "negative_impls": false,
      "associated_item_constraints": false,
      "coupons": false
    }
  },
  "linter_configuration": "Off",
  "plugins": {}
}

{
  "name": "b",
  "source_paths": [
    "[ROOT]/b/src/lib.cairo"
  ],
  "settings": {
    "name": "b",
    "edition": "2024_07",
    "version": "0.1.0",
    "cfg_set": [
      [
        "target",
        "lib"
      ],
      [
        "target",
        "test"
      ],
      "test"
    ],
    "dependencies": {
      "b": {
        "discriminator": "b 0.1.0 (path+[ROOT_URL]b/Scarb.toml)"
      },
      "core": {
        "discriminator": null
      }
    },
    "experimental_features": {
      "negative_impls": false,
      "associated_item_constraints": false,
      "coupons": false
    }
  },
  "linter_configuration": "Off",
  "plugins": {}
}

{
  "name": "core",
  "source_paths": [
    "[SCARB_REGISTRY_STD]/core/src/lib.cairo"
  ],
  "settings": {
    "name": "core",
    "edition": "2024_07",
    "version": "2.11.4",
    "cfg_set": [
      [
        "target",
        "lib"
      ],
      [
        "target",
        "test"
      ]
    ],
    "dependencies": {
      "core": {
        "discriminator": null
      }
    },
    "experimental_features": {
      "negative_impls": true,
      "associated_item_constraints": true,
      "coupons": true
    }
  },
  "linter_configuration": "Off",
  "plugins": {
    "builtin_plugins": [
      "Executable",
      "CairoTest"
    ]
  }
}
```'''
analyzed_crates_diff = '''
@@ -47,32 +47,31 @@
   "source_paths": [
     "[ROOT]/b/src/lib.cairo"
   ],
   "settings": {
     "name": "b",
     "edition": "2024_07",
     "version": "0.1.0",
     "cfg_set": [
       [
         "target",
         "lib"
       ],
       [
         "target",
         "test"
-      ],
-      "test"
+      ]
     ],
     "dependencies": {
       "b": {
         "discriminator": "b 0.1.0 (path+[ROOT_URL]b/Scarb.toml)"
       },
       "core": {
         "discriminator": null
       }
     },
     "experimental_features": {
       "negative_impls": false,
       "associated_item_constraints": false,
       "coupons": false
     }
   },
'''
